repository-s3/src/main/java/io/aiven/elasticsearch/repositories/s3/S3RepositoryStorageIOProvider.java

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright 2020 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.elasticsearch.repositories.s3;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import io.aiven.elasticsearch.repositories.Permissions;
+import io.aiven.elasticsearch.repositories.RepositoryStorageIOProvider;
+import io.aiven.elasticsearch.repositories.io.CryptoIOProvider;
+import io.aiven.elasticsearch.repositories.security.EncryptionKeyProvider;
+
+import com.amazonaws.AmazonClientException;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.DeleteObjectsRequest;
+import com.amazonaws.services.s3.model.ListObjectsV2Request;
+import com.amazonaws.services.s3.model.ListObjectsV2Result;
+import com.amazonaws.services.s3.model.MultiObjectDeleteException;
+import com.amazonaws.services.s3.model.S3ObjectInputStream;
+import org.elasticsearch.common.blobstore.BlobStoreException;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.ByteSizeUnit;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class S3RepositoryStorageIOProvider extends RepositoryStorageIOProvider<AmazonS3> {
+
+    static final Setting<String> BUCKET_NAME =
+            Setting.simpleString(
+                    "bucket_name",
+                    Setting.Property.NodeScope,
+                    Setting.Property.Dynamic);
+
+    static final Setting<ByteSizeValue> MULTIPART_UPLOAD_PART_SIZE =
+            Setting.byteSizeSetting(
+                    "multipart_upload_part_size",
+                    new ByteSizeValue(10, ByteSizeUnit.MB),
+                    new ByteSizeValue(5, ByteSizeUnit.MB),
+                    new ByteSizeValue(100, ByteSizeUnit.MB),
```

Comment:
Any reason to not let it go up to 5 GB (S3's max)?

Reply:
Do you mean maximum value?

Reply:
Yeah